File: plugins/alignments/src/shared/pairOrientation.ts
// Pair orientation string (e.g. F1R2, R2F1) for a read pair.
//
// Must be computed identically from either mate, otherwise the two mates of one
// normal pair disagree and get colored as different orientations (e.g. one gray
// LR, one teal RL). The "left" mate is therefore chosen by a total order on
// (refId, position) that both mates evaluate the same way, with a read1-first
// tie-break for equal loci. When the mate position is unknown the read1-first
// rule alone keeps the two mates consistent.
export function getPairOrientation({
  isRead1,
  isSelfRev,
  isMateRev,
  selfRefId,
  selfPos,
  mateRefId,
  matePos,
}: {
  isRead1: boolean
  isSelfRev: boolean
  isMateRev: boolean
  selfRefId: number
  selfPos: number
  mateRefId: number | undefined
  matePos: number | undefined
}) {
  const selfStrand = isSelfRev ? 'R' : 'F'
  const mateStrand = isMateRev ? 'R' : 'F'
  const selfNum = isRead1 ? '1' : '2'
  const mateNum = isRead1 ? '2' : '1'

  const selfIsLeft =
    mateRefId === undefined || matePos === undefined
      ? isRead1
      : selfRefId !== mateRefId
        ? selfRefId < mateRefId
        : selfPos !== matePos
          ? selfPos < matePos
          : isRead1

  return selfIsLeft
    ? selfStrand + selfNum + mateStrand + mateNum
    : mateStrand + mateNum + selfStrand + selfNum
}

File: plugins/linear-genome-view/src/LinearGenomeView/rhelpers/pair_orientation.R
# Classify a read pair's orientation into the IGV FR-library categories
# (LR normal / RL / RR / LL), reproducing JBrowse's getPairOrientation +
# pairDirection. Vectorized over a BAM's flag / position columns. The leftmost
# mate (self_left) is chosen consistently from either read - by position within a
# chromosome, read1-first when positions tie or the mate position is unknown - so
# the two mates of one pair always agree. Returns NA for unpaired reads or
# orientations outside the FR set. self_pos/mate_pos are 1-based leftmost coords.
pair_orientation <- function(flag, self_pos, mate_pos) {
  is_paired <- bitwAnd(flag, 0x1L) != 0
  is_read1  <- bitwAnd(flag, 0x40L) != 0
  self_str  <- ifelse(bitwAnd(flag, 0x10L) != 0, "R", "F")
  mate_str  <- ifelse(bitwAnd(flag, 0x20L) != 0, "R", "F")
  self_num  <- ifelse(is_read1, "1", "2")
  mate_num  <- ifelse(is_read1, "2", "1")
  known     <- !is.na(mate_pos) & mate_pos > 0
  self_left <- ifelse(!known | self_pos == mate_pos, is_read1, self_pos < mate_pos)
  o <- ifelse(self_left,
    paste0(self_str, self_num, mate_str, mate_num),
    paste0(mate_str, mate_num, self_str, self_num))
  fr <- c(F1R2 = "LR", F2R1 = "LR", R1F2 = "RL", R2F1 = "RL",
          R1R2 = "RR", R2R1 = "RR", F1F2 = "LL", F2F1 = "LL")
  ifelse(is_paired, unname(fr[o]), NA_character_)
}
